src/Plugin/Commerce/PaymentGateway/OnsiteBase.php

```diff
@@ -613,4 +614,70 @@ protected function orderIsInitialSubscription(OrderInterface $order) {
     return FALSE;
   }
 
+   /**
+   * Sets the remote customer ID for the given user.
+   *
+   * @param \Drupal\user\UserInterface $account
+   *   The user account.
+   * @param string $remote_id
+   *   The remote customer ID.
+   */
+  protected function setRemoteCustomerId(UserInterface $account, $remote_id) {
+    if ($account->isAuthenticated()) {
+      /** @var \Drupal\commerce\Plugin\Field\FieldType\RemoteIdFieldItemListInterface $remote_ids */
+      $remote_ids = $account->get('commerce_remote_id');
+      $remote_ids->setByProvider(
+        $this->remoteCustomerProviderKey(),
+        $remote_id
+      );
+    }
+  }
+
+  /**
+   * Gets the remote customer ID for the given user.
+   *
+   * The remote customer ID is specific to a payment gateway instance
+   * in the configured mode. This allows the gateway to skip test customers
+   * after the gateway has been switched to live mode.
+   *
+   * @param \Drupal\user\UserInterface $account
+   *   The user account.
+   *
+   * @return string
+   *   The remote customer ID, or NULL if none found.
+   */
+  protected function getRemoteCustomerId(UserInterface $account) {
+    $remote_id = NULL;
+    if ($account->isAuthenticated()) {
+      $key = $this->remoteCustomerProviderKey();
+
+      /** @var \Drupal\commerce\Plugin\Field\FieldType\RemoteIdFieldItemListInterface $remote_ids */
+      $remote_ids = $account->get('commerce_remote_id');
+      $remote_id = $remote_ids->getByProvider($key);
+      // Gateways used to key customer IDs by module name, migrate that data.
+      if (!$remote_id) {
+        $remote_id = $remote_ids->getByProvider($this->pluginDefinition['provider']);
+        if ($remote_id) {
+          $remote_ids->setByProvider($this->pluginDefinition['provider'], NULL);
+          $remote_ids->setByProvider(
+            $key,
+            $remote_id
+          );
+          $account->save();
+        }
+      }
+    }
+    return $remote_id;
+  }
+
+  /**
+   * Returns the provider key to be used for setting remote ID.
+   *
+   * @return string
+   *   The provider key string for setting remote ID.
+   */
+  protected function remoteCustomerProviderKey() {
+    return 'bluesnap|' . $this->configuration['username'] . '|'. $this->getMode();
```

Comment:
So, basically, we are associating the remote customer ID NOT per payment gateway, but per the gateway username used in the credentials. So, multiple payment gateways (like ACH/Hosted Payment fields) would have the same credentials and the customer would only have one vaulted shopper ID created per multiple payment gateways.

Looks good to me.

Reply:
Yep. commerce payment module uses payment gateway entity ID here, we are using bluesnap username instead